xla/stream_executor/command_buffer.h

```diff
@@ -172,177 +127,126 @@ class CommandBuffer {
-  //     body_builder()
-  //     cond_builder()
-  //
-  // We use execution scope builder for the condition because we have to build
-  // condition twice: (1) before the conditional node in the scope defined by
-  // `execution_scope_id` (2) inside the loop body with default execution scope.
-  virtual absl::Status While(ExecutionScopeId execution_scope_id,
-                             DeviceMemory<bool> pred,
-                             ExecutionScopeBuilder cond_builder,
-                             Builder body_builder) = 0;
-
-  // Adds a conditional While operation to default execution scope.
-  absl::Status While(DeviceMemory<bool> pred,
-                     ExecutionScopeBuilder cond_builder, Builder body_builder) {
-    return While(kDefaultExecutionScope, pred, cond_builder, body_builder);
-  }
+  // Create a new conditional handle in the underlying
+  // graph.
+  virtual absl::StatusOr<GraphConditionalHandle> CreateConditionalHandle() = 0;
+
+  // Adds a new conditional node to the graph and creates a
+  // corresponding nested command buffer.
+  virtual absl::StatusOr<ConditionalNodeResult> CreateConditionalNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle conditional,
+      ConditionType type) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetIfElseConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle then_condition,
+      GraphConditionalHandle else_condition, DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetIfElseConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle then_condition,
+      GraphConditionalHandle else_condition, DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetIfConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle then_condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetIfConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle then_condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetForConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle condition,
+      DeviceMemory<int32_t> loop_counter, int32_t iterations) = 0;
+
+  virtual absl::Status UpdateSetForConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle condition,
+      DeviceMemory<int32_t> loop_counter, int32_t iterations) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetWhileConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetWhileConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetCaseConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle handle0,
```

Comment:
take conditional handles as absl::Span?

Reply:
dependencies are mostly passed by rvalue in the code, so we should keep as std::vector?  there is no extra copy if passing with rvalue ?

Reply:
absl::Span is more consistent with XLA codebase

Reply:
OK, let me fix that.